lib/ProductOpener/APIProductServices.pm

```diff
@@ -0,0 +1,218 @@
+
+=head3 Response
+
+The response is in the JSON API v3 response format, with a resulting product object.
+
+=cut
+
+package ProductOpener::APIProductServices;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&product_services_api
+	);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Users qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::API qw/:all/;
+
+use Encode;
+
+=head2 echo_service ($product_ref)
+
+Echo service that returns the input product unchanged.
+
+=cut
+
+sub echo_service ($product_ref, $updated_product_fields_ref) {
+
+	return;
+}
+
+my %service_functions = (
+	echo => \&echo_service,
+	estimate_ingredients_percent => \&ProductOpener::Ingredients::estimate_ingredients_percent_service,
+	analyze_ingredients => \&ProductOpener::Ingredients::analyze_ingredients_service,
+);
+
+=head2 product_services_api()
+
+Process API v3 product services requests.
+
+=cut
+
+sub product_services_api ($request_ref) {
+
+	$log->debug("product_services_api - start", {request => $request_ref}) if $log->is_debug();
+
+	my $response_ref = $request_ref->{api_response};
+	my $request_body_ref = $request_ref->{body_json};
+
+	$log->debug("product_services_api - body", {request_body => $request_body_ref}) if $log->is_debug();
+
+	my $error = 0;
+
+	# Check that we have an input body
+	if (not defined $request_body_ref) {
+		$log->error("product_services_api - missing or invalid input body", {}) if $log->is_error();
+		$error = 1;
+	}
+	else {
+		# Check that we have the input body fields we expect
+
+		if (not defined $request_body_ref->{product}) {
+			$log->error("product_services_api - missing input product", {request_body => $request_body_ref})
+				if $log->is_error();
+			add_error(
+				$response_ref,
+				{
+					message => {id => "missing_field"},
+					field => {id => "product"},
+					impact => {id => "failure"},
+				}
+			);
+			$error = 1;
+		}
+
+		if (not defined $request_body_ref->{services}) {
+			$log->error("product_services_api - missing services", {request_body => $request_body_ref})
+				if $log->is_error();
+			add_error(
+				$response_ref,
+				{
+					message => {id => "missing_field"},
+					field => {id => "services"},
+					impact => {id => "failure"},
+				}
+			);
+			$error = 1;
+		}
+		elsif (ref($request_body_ref->{services}) ne 'ARRAY') {
+			add_error(
+				$response_ref,
+				{
+					message => {id => "invalid_type_must_be_array"},
+					field => {id => "services"},
+					impact => {id => "failure"},
+				}
+			);
+			$error = 1;
+		}
+		else {
+			# Echo back the services that were requested
+			$response_ref->{services} = $request_body_ref->{services};
+		}
+	}
```

Comment:
My preferences would be that when we have meaningful blocks like that (check inputs), we put them in a separate private sub. This helps reading the code.